### PR TITLE
SW-5466: Support Velodyne and other point types in ouster-ros driver

### LINKED ISSUE #15: Change the format into velodyne format
Hi everyone:

I am sorry that I am fresh and I would like to ask that how can I change the format of ouster into Velodyne like format. I found the code here:
<html>
<body>

void ouster_driver::OS1::convert2XYZIR | ( | const CloudOS1 & | in,
  |   | CloudOS1XYZIR & | out
  | )

but I did not know which code and which part of the code should I change, thx!

<br class="Apple-interchange-newline">
</body>
</html>void ouster_driver::OS1::convert2XYZIR	(	const [CloudOS1](http://docs.ros.org/en/melodic/api/ouster_driver/html/namespaceouster__driver_1_1OS1.html#a9455ea14ce69167cfdc9a34150a1fa31) & 	in,
[CloudOS1XYZIR](http://docs.ros.org/en/melodic/api/ouster_driver/html/namespaceouster__driver_1_1OS1.html#ac346ac2772c984d044b5d6c3d808b3fa) & 	out 
)

### LINKED ISSUE #29: Parameter for Controlling PCL PointType
Hello,

Thanks for the Ouster driver, it's works well! I would like to make one request, which is to my knowlesge, the user does not have any control over what fields the Ouster publishes. The driver will always publish X, Y, Z, intensity, t, reflectivity, ring, noise, and range. This is a lot of data that to most users is not necessary. In my specific application, I want XYZI because I am fusing the point cloud with XYZI point clouds from other depth sensors onboard a robot. Therefore, having the same PointType is convenient. I can envision other folks wanting to disable fields they don't need to reduce the amount of computational/communications resources required to run the driver.

In order to resolve this issue, I had to make changes to the driver source code (ouster_ros/include/ouster_ros/point_os1.h), which I would prefer to not do.

Hacky solution: commented out unwanted PCL fields

Thanks!

Gene

### LINKED ISSUE #210: Add low data rate point type
**Is your feature request related to a problem? Please describe.**
I would like the pointclouds to use less bandwidth when I use the [low data rate profile](https://static.ouster.dev/sensor-docs/image_route1/image_route2/sensor_data/sensor-data.html#low-data-rate-profile) from the sensor.

**Describe the solution you'd like**
When the low data rate profile is selected, use a different point type that has lower precision and fewer fields. For example:

```cpp
struct EIGEN_ALIGN16 LowDataRatePoint {
    PCL_ADD_POINT4D;
    uint32_t t;
    uint8_t reflectivity;
    uint16_t ring;
    uint8_t ambient;
    uint16_t range;
    EIGEN_MAKE_ALIGNED_OPERATOR_NEW
};
```

**Describe alternatives you've considered**
Not sure.

**Targeted Platform (please complete the following information only if applicable, otherwise dot N/A):**
- Ouster Sensor? Rev 6 and Rev 7
- Ouster Firmware Version? 2.3+
- ROS version/distro? N/A
- Operating System? N/A
- Machine Architecture? N/A

### PR DESCRIPTION
**Pull request by Samahu**
## Related Issues & PRs
- resolves #15 
- resolves #29
- resolves #210 
- related #97
- related #217
- related #223

## Summary of Changes
- added the ability to customize the published point clouds(s) to velodyne point cloud format and
  other common pcl point types.
- made ouster_image_component operate separately from ouster_cloud_component

## Validation
- Run the driver with default parameters and verify no observable behavioral changes.
- Change `point_type` parameter to `native` and verify that it reflects the same point cloud format that we get from the sensor.
- Change `point_type` parameter to `xyz` and verify that the published point cloud format only has `xyz` fields, regardless of the underlying sensor format.
- Change `point_type` parameter to `xyzi` and verify that the published point cloud format only has `xyz` fields plus intensity, regardless of the underlying sensor format (you should get a warning when the low data mode is selected for this mode).
- Change `point_type` parameter to `xyzir` and verify that the published point cloud format only has `xyz` fields plus intensity plus ring, regardless of the underlying sensor format (you should get a warning when the low data mode is selected for this mode).

### CODE REVIEW

File: ouster-ros/config/driver_params.yaml [hunk 1]
```diff
@@ -67,3 +67,18 @@ ouster/os_driver:
...
+     # point_type[optional]: choose from: {original, native, xyz, xyzi, xyzir}
+     # Here is a breif description of each option:
+     #  - original: ouster_ros::Point
+     #  - native: directly maps all fields as published by the sensor to an
+     #          equivalent point cloud representation with the additon of ring
+     #          and timestamp fields.
+     #  - xyz: the simplest point type, only has {x, y, z}
+     #  - xyzi: same as xyz point type but adds intensity(signal) field. this
+     #          type is not compatible with the low data profile.
+     #  - xyzi: same as xyzi type but adds ring(channel) field.
```
Comment (mtswisher) [inline ouster-ros/config/driver_params.yaml:79]: xyzi should be xyzir on this line

File: ouster-ros/include/ouster_ros/sensor_point_types.h [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+ *
+ * @file sensor_point_types.h
+ * @brief native sensor point types
+ * The following are one-to-one mapping of pcl point representatios that could
+ * fit the data sent by the sensor (with the addition of t and ring fields),
+ * All of these representation follow the same order of LaserScan fields:
+ * 1. range
+ * 2. signal
+ * 3. refelctivity
+ * 4. nea_ir
```
Comment (mtswisher) [inline ouster-ros/include/ouster_ros/sensor_point_types.h:13]: near_ir :D

File: ouster-ros/include/ouster_ros/sensor_point_types.h [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+ * The only Point_RNG15_RFL8_NIR8 aks LOW_DATA which does not omits the signal
```
Comment (mtswisher) [inline ouster-ros/include/ouster_ros/sensor_point_types.h:14]: aka as opposed to aks

**Review by mtswisher — CHANGES_REQUESTED** [review]
Just made some typo changes :)  How do you decide whether to designate a function as inline? I haven't seen the factory pattern in quite awhile so it was cool to see that solve this problem. Will probably need to dig into some more to really understand what is going on but LGTM :)

**Comment by Samahu** [comment]
> Just made some typo changes :)

Addressed!

> How do you decide whether to designate a function as inline?

Typically for short one liner methods this could minimally increase the code size but avoid having to invoke the method which is always faster .. With modern compilers the **inline** keyword only serves as a suggestion, compilers could inline methods automatically .. and could also ignore the suggestion if the method body is too long (in this case one needs to use **__forceinline**).  
 
> I haven't seen the factory pattern in quite awhile so it was cool to see that solve this problem. Will probably need to dig into some more to really understand what is going on but LGTM :)

Yeah, the factory pattern is definitely needed here. I will add more comments to the PR to explain what's going on and how could developers use existing code to implement other point types that are not included here.

**Review by twslankard — APPROVED** [review]

**Pull request merged**